README.md
# SwordieDB
### [SwordieDB PyPi Page](https://pypi.org/project/swordiedb/)

[![Downloads](https://static.pepy.tech/personalized-badge/swordiedb?period=total&units=international_system&left_color=black&right_color=blue&left_text=Total%20Downloads)](https://pepy.tech/project/swordiedb) [![Downloads](https://static.pepy.tech/personalized-badge/swordiedb?period=month&units=international_system&left_color=black&right_color=blue&left_text=Monthly%20Downloads)](https://pepy.tech/project/swordiedb) [![Downloads](https://static.pepy.tech/personalized-badge/lazuli?period=week&units=international_system&left_color=black&right_color=blue&left_text=Weekly%20Downloads)](https://pepy.tech/project/swordiedb)  

SwordieDB is a pip-compatible Python package designed for use in development of SwordieMS-based MapleStory private server tools (e.g. Discord bots).  

The SwordieDB API allows access to character stats in the form of objects, for easy manipulation.  

#### **Current Status:** Fully Released
**Overall Developmental Progress:** ![](https://progress-bar.dev/100) 
  
**Character Handling:** ![](https://progress-bar.dev/100)    
**User Handling:** ![](https://progress-bar.dev/100)    
**Inventory Handling:** ![](https://progress-bar.dev/100)    
## Installation
Note: You are recommended to install to the project virtual environment.
- In the terminal type the follow command: `pip install swordiedb`  
    - Note: use system shell and not the Python REPL for pip!
    - To avoid any issues make sure to have these commands run prior to installing SwordieDB:
        - `pip install -U setuptools`
        - `pip install -U wheel`
- Now to verify that you have access to the swordie_db module:
    - Try this import statement in the Python terminal: `from swordie_db.database import SwordieDB`
- Then you are done! Refer to [examples](https://github.com/Bratah123/SwordieDB/wiki/Sample-Code-Fragments) on how to use them. :octocat:
## Examples
See the [Code Examples page](https://github.com/Bratah123/SwordieDB/wiki/Sample-Code-Fragments) for sample code fragments.
## Technical Details
Please refer to the [Technical Details page](https://github.com/Bratah123/SwordieDB/wiki/Technical-Details) for details, as well as detailed instructions for how to build this package for testing.  
## API Docs
The API Docs have been split out into a sister project [here](https://kookiiestudios.github.io/ProjectShieldieDB/).

## Suggestions
Suggestions for new features/functionality can be written as an issue from the ["Issues"](https://github.com/Bratah123/SwordieDB/issues) tab on github.

## Credits
- Credits to the original swordie devs!


CHANGELOG.txt
Change Log
==========

1.0.2 (1/20/2021)
-----------------
- Added more attributes to user

1.0.1 (11/15/2020)
------------------
- Added get_gender_name() to character class

1.0.0 (11/13/2020)
------------------
- Official release of SwordieDB API!
- All unit testing has been done to determine full 100% functioning methods and classes (Kookiie)

0.2.3 (11/12/2020)
------------------
- Added more functionality to Inventory Class
- This should complete all important modules of this library (awaiting testing)

0.2.2 (11/12/2020)
------------------
- Added setup inventory support
- Added more examples on how to use inventory class

0.2.1 (11/12/2020)
------------------
- Refactor to how database calls are used in character class
- Added inventory handling (able to see character's inventory)

0.2.0 (11/12/2020)
------------------
- Raised minor version to 2 as all of User and Character functionality has been tested to work!
- Updated documentation to Google Conventions (Kookiie)

0.1.4 (11/12/2020)
------------------
- Added nxPrepaid support to user
- Updated docstrings
- Added inventory id support to character
- User at 100% completion

0.1.3 (11/12/2020)
------------------
- Fixed an accidental recursive call of donation_points setter
- Unit testing for User Class is completed with 100% working functions
- Updated docstring to Google conventions
- Changelog is now reversed

0.1.2 (11/11/2020)
------------------
- Fixed User Class instantiation

0.1.1 (11/11/2020)
------------------
- Added a new "User" Class
- Added new helper methods to existing classes
- Fixed some functions not using custom port in SwordieDB
- Documentation addition in github wiki

0.1.0 (11/11/2020)
------------------
- Character Handling is now 100% working and tested
- Removed unnecessary imports
- Fixed missing Job IDs
- Minor Documentation Updates

0.0.6 (11/10/2020)
------------------
- Fixed getting a job's name from the job id
- Minor Documentation tweaks

0.0.5 (11/10/2020)
------------------
- Added a port parameter to SwordieDB Constructor

0.0.4 (11/10/2020)
------------------
- fixed get_job_name()

0.0.3 (11/10/2020)
------------------
- Minor documentation edit

0.0.2 (11/10/2020
-----------------
- Fixed Typos

0.0.1 (11/10/2020)
------------------
- First Release
- Completed Character Object

KOOKIIE_TEST.py
"""This is a unit test for checking basic functionality

This can be run directly for a quick test of functionality (only supports KOOKIIE's DB at the moment)
Copyright KOOKIIE Studios 2020. All rights reserved.
"""
from swordie_db.database import SwordieDB

print("This is a unit test for checking basic functionality.")
print("If no error messages are shown, it means that the program is working correctly.")
print("----------------------------------\n")

# Import DB
try:
	spirit = SwordieDB(host="127.0.0.1", password="", user="root", schema="spirit", port=3306)
except Exception as e:
	print(f"Error has occurred whist attempting to load DB: \n{e}")
	sys.exit(0)

# Character info fetching tests
char = spirit.get_char_by_name("Kookiiee")
if char == None:
	print("CRITICAL ERROR: UNABLE TO FETCH CHARACTER BY NAME! TERMINATING...")
	sys.exit(0)
user_id = spirit.get_user_id_by_name("Kookiiee")
if user_id == None:
	print("CRITICAL ERROR: UNABLE TO FETCH USER ID BY NAME! TERMINATING...")
	sys.exit(0)

# Fetch character object properties
name = char.name  # getter for name; Sting
meso = char.money  # getter for mesos; String
fame = char.fame  # getter for fame; Int
job = char.get_job_name()  # return job name from ID via Hashmap; String
job_id = char.job  # getter for job id; Int
level = char.level  # getter for levels; Int
honor_exp = char.get_stat_by_column("honorexp")  # Int

map = char.map  # getter for map ID; String
face = char.face  # getter for face ID; Int
hair = char.hair  # getter for hair ID; Int
skin = char.skin  # getter for skin ID; Int
exp = char.exp  # getter for EXP amount; String

strength = char.strength  # getter for STR; Int
dex = char.dex  # getter for DEX; Int
inte = char.inte  # getter for INT; Int
luk = char.luk  # getter for LUK; Int
primary_stats = char.get_primary_stats()  # returns a dictionary of the 4 main stats; dictionary
hp = char.max_hp  # getter for max HP; Int
mp = char.max_mp  # getter for max MP; Int
ap = char.ap  # getter for free AP; Int
sp = char.sp  # getter for free SP; Int


print("Now checking char info from DB...")

print("Checking the ability to fetch the following information from DB: ")
print("    > Character Name; Mesos; Fame; Job name; Job ID; Character Level; Honour EXP. \n")
assert name == "Kookiiee", f"Critical Error: Name test failed! Name: {name}; Type: {type(name)}"
assert meso == "0", f"Meso test failed! Meso count: {meso}; Type: {type(meso)}"
assert fame == 0, f"Fame test failed! Fame count: {fame}; Type: {type(fame)}"
assert job == "Beginner", f"Job name test failed! Job name: {job}; Type: {type(job)}"
assert job_id == 0, f"Job ID test failed! Job ID: {job_id}; Type: {type(job_id)}"
assert level == 1, f"Character level test failed! Level count: {level}; Type: {type(level)}"
assert honor_exp == 0, f"Honour EXP test failed! Honour count: {honor_exp}; Type: {type(honor_exp)}"

print("Checking the ability to fetch the following information from DB: ")
print("    > Map; Face; Hair; Skin; EXP. \n")
assert map == "4000011", f"Map ID test failed! Map ID: {map}; Type: {type(map)}"
assert face == 23300, f"Face ID test failed! Face ID: {face}; Type: {type(face)}"
assert hair == 36786, f"Hair ID test failed! Hair ID: {hair}; Type: {type(hair)}"
assert skin == 0, f"Skin ID test failed! Skin ID: {skin}; Type: {type(skin)}"
assert exp == "0", f"EXP test failed! EXP amount: {exp}; Type: {type(exp)}"

print("Checking the ability to fetch the following information from DB: ")
print("    > STR; DEX; INT; LUK; Primary Stats; HP; MP; AP; SP. \n")
assert strength == 12, f"STR test failed! STR amount: {strength}; Type: {type(strength)}"
assert dex == 5, f"DEX test failed! DEX amount: {dex}; Type: {type(dex)}"
assert inte == 4, f"INT test failed! INT amount: {inte}; Type: {type(inte)}"
assert luk == 4, f"LUK test failed! LUK amount: {luk}; Type: {type(luk)}"
assert primary_stats == {'str': 12, 'dex': 5, 'int': 4, 'luk': 4}, f"Primary Stats test failed! \nExpected: {{'str': 12, 'dex': 5, 'int': 4, 'luk': 4}} \nEncountered: {primary_stats}"
assert hp == 50, f"HP test failed! HP amount: {hp}; Type: {type(hp)}"
assert mp == 0, f"MP test failed! MP amount: {mp}; Type: {type(mp)}"
assert ap == 0, f"AP test failed! AP amount: {ap}; Type: {type(ap)}"
assert sp == 0, f"SP test failed! SP amount: {sp}; Type: {type(sp)}"

print("Checking the ability to fetch the following information from DB: ")
print("    > User ID. \n")
assert user_id == 5, f"Database method 'get_user_id_by_name' test failed! User ID: {user_id}; Type: {type(user_id)}"
print("Char info fetch tests complete!")
print("----------------------------------")

# Character info setting tests
print("Checking the ability to write the following information to DB: ")
print("    > Mesos; Fame. \n")
# Set values for Meso and Fame
char.money = "314159" # Sets money to 314,159 mesos in the database
char.fame = 3  # Sets fame to 3 in the database

meso = char.money  # re-fetch data
assert meso == "314159", f"Meso setting test failed! Meso count: {meso}; Type: {type(meso)}"
char.add_mesos(2827433)  # Adds 2,827,433 to the current meso count, and saves to DB
# character now has 3,141,592 mesos
meso = char.money  # re-fetch data
assert meso == "3141592", f"Meso adding test failed! Meso count: {meso}; Type: {type(meso)}"
char.money = "0"  # reset to baseline

fame = char.fame  # re-fetch data
assert fame == 3, f"Fame setting test failed! Fame count: {fame}; Type: {type(fame)}"
char.add_fame(28) # Adds 28 fame to the existing count and saves to database
# character fame is now 31
fame = char.fame  # re-fetch data
assert fame == 31, f"Fame adding test failed! Fame count: {fame}; Type: {type(fame)}"
char.fame = 0  # reset to baseline


print("\nChecking the ability to write the following information to DB: ")
print("    > Level; Job ID; Name; Map. \n")
# Set values for Level, Job ID, Name, and Map
char.level = 10
char.job = 100  # set job ID to warrior
char.name = "Kookiie"
char.map = "100000000"

level = char.level  # re-fetch data
assert level == 10, f"Character level setting test failed! Level count: {level}; Type: {type(level)}"
char.add_level(21)  # Adds 21 to the existing count and saves to database
# character is now level 31
level = char.level  # re-fetch data
assert level == 31, f"Character level adding test failed! Level count: {level}; Type: {type(level)}"
char.level = 1  # reset to baseline

job_id = char.job  # re-fetch data
assert job_id == 100, f"Job ID setting test failed! Job ID: {job_id}; Type: {type(job_id)}"
char.job = 0  # reset job ID to beginner

name = char.name  # re-fetch data
assert name == "Kookiie", f"Name setting test failed! Name: {name}; Type: {type(name)}"
char.name = "Kookiiee"  # reset to baseline

map = char.map  # re-fetch data
assert map == "100000000", f"Map ID setting test failed! Map ID: {map}; Type: {type(map)}"
char.map = 4000011  # reset to baseline


print("\nChecking the ability to write the following information to DB: ")
print("    > Face; Hair; Skin; EXP. \n")
# Set values for Face, Hair, Skin, EXP
char.face = 20010
char.hair = 30027
char.skin = 2
char.exp = "314159"

face = char.face  # re-fetch data
assert face == 20010, f"Face ID setting test failed! Face ID: {face}; Type: {type(face)}"
char.face = 23300  # reset to baseline

hair = char.hair  # re-fetch data
assert hair == 30027, f"Hair ID setting test failed! Hair ID: {hair}; Type: {type(hair)}"
char.hair = 36786  # reset to baseline

skin = char.skin  # re-fetch data
assert skin == 2, f"Skin ID setting test failed! Skin ID: {skin}; Type: {type(skin)}"
char.skin = 0  # reset to baseline

exp = char.exp  # re-fetch data
assert exp == "314159", f"EXP test setting failed! EXP amount: {exp}; Type: {type(exp)}"
char.add_exp(2827433)
exp = char.exp  # re-fetch data
assert exp == "3141592", f"EXP test adding failed! EXP amount: {exp}; Type: {type(exp)}"
char.exp = "0"  # reset to baseline


print("\nChecking the ability to write the following information to DB: ")
print("    > STR; DEX; INT; LUK. \n")
# Set values for STR, DEX, INT, LUK
char.strength = 31
char.dex = 31
char.inte = 31
char.luk = 31

strength = char.strength  # re-fetch data
assert strength == 31, f"STR setting test failed! STR amount: {strength}; Type: {type(strength)}"
char.add_str(1)
strength = char.strength  # re-fetch data
assert strength == 32, f"STR adding test failed! STR amount: {strength}; Type: {type(strength)}"
char.strength = 12  # reset to baseline

dex = char.dex  # re-fetch data
assert dex == 31, f"DEX setting test failed! DEX amount: {dex}; Type: {type(dex)}"
char.add_dex(1)
dex = char.dex  # re-fetch data
assert dex == 32, f"DEX adding test failed! DEX amount: {dex}; Type: {type(dex)}"
char.dex = 5  # reset to baseline

inte = char.inte  # re-fetch data
assert inte == 31, f"INT setting test failed! INT amount: {inte}; Type: {type(inte)}"
char.add_inte(1)
inte = char.inte  # re-fetch data
assert inte == 32, f"INT adding test failed! INT amount: {inte}; Type: {type(inte)}"
char.inte = 4  # reset to baseline

luk = char.luk  # re-fetch data
assert luk == 31, f"LUK setting test failed! LUK amount: {luk}; Type: {type(luk)}"
char.add_luk(1)
luk = char.luk  # re-fetch data
assert luk == 32, f"LUK adding test failed! LUK amount: {luk}; Type: {type(luk)}"
char.luk = 4  # reset to baseline


print("\nChecking the ability to write the following information to DB: ")
print("    > HP; MP; AP; SP. \n")
# Set values for HP, MP, AP, SP
char.max_hp = 31
char.max_mp = 31
char.ap = 31
char.sp = 31

hp = char.max_hp  # re-fetch data
assert hp == 31, f"HP setting test failed! HP amount: {hp}; Type: {type(hp)}"
char.add_max_hp(1)
hp = char.max_hp  # re-fetch data
assert hp == 32, f"HP adding test failed! HP amount: {hp}; Type: {type(hp)}"
char.max_hp = 50  # reset to baseline

mp = char.max_mp  # re-fetch data
assert mp == 31, f"MP setting test failed! MP amount: {mp}; Type: {type(mp)}"
char.add_max_mp(1)
mp = char.max_mp  # re-fetch data
assert mp == 32, f"MP adding test failed! MP amount: {mp}; Type: {type(mp)}"
char.max_mp = 0  # reset to baseline

ap = char.ap  # re-fetch data
assert ap == 31, f"AP setting test failed! AP amount: {ap}; Type: {type(ap)}"
char.add_ap(1)
ap = char.ap  # re-fetch data
assert ap == 32, f"AP adding test failed! AP amount: {ap}; Type: {type(ap)}"
char.ap = 0  # reset to baseline

sp = char.sp  # re-fetch data
assert sp == 31, f"SP setting test failed! SP amount: {sp}; Type: {type(sp)}"
char.add_sp(1)
sp = char.sp  # re-fetch data
assert sp == 32, f"SP adding test failed! SP amount: {sp}; Type: {type(sp)}"
char.sp = 0  # reset to baseline

print("\nChar info write tests complete!")
print("----------------------------------")


example_use.py
from swordie_db.database import SwordieDB
import time

swordie_data = SwordieDB(schema="spirit")  # If no parameters are given, it will attempt to connect to localhost

char = swordie_data.get_char_by_name("brandon")

assert char is not None, "Character does not exist"

print("Character Name:", char.name)
print("Character Level:", char.level)
print("Job:", char.get_job_name())
print("Mesos:", char.money)
print("Fame:", char.fame)
print("Character image", char.get_char_img())

primary_stats = char.get_primary_stats()

for stat in primary_stats:
    print(f"{stat}: {primary_stats[stat]}")
    # Str
    # Dex
    # Int
    # Luk

# User Class

user = char.user
user.is_admin()

# Inventory

print("Consume Inventory:", char.inventory.consume_inv)

equip_inv = char.inventory.consume_inv
print("Testing equip checks....")


def test_check_speed():
    for bag_index in equip_inv:
        if 2000000 in equip_inv[bag_index].values():
            return True
    return False


def test_check_speed_equals():
    for bag_index in equip_inv:
        if 2000000 == equip_inv[bag_index]['itemid']:
            return True
    return False


print("Starting list comprehension test")

start_time = time.time()

for i in range(100000):
    test_check_speed()

end_time = time.time()

print(f"Completed List Comprehension 100000 iterations in: {end_time - start_time} seconds")

print("Starting == test")

start_time1 = time.time()

for n in range(100000):
    test_check_speed_equals()

end_time1 = time.time()

print(f"Completed 100000 iterations of == operations in: {end_time1 - start_time1} seconds")


swordie_db/user.py
"""This module holds the User class for the SwordieDB package.

Copyright 2020 TEAM SPIRIT. All rights reserved.
Use of this source code is governed by a MIT-style license that can be found in the LICENSE file.
Refer to database.py or the project wiki on GitHub for usage examples.
"""
import mysql.connector as con


class User:
    """User object; models SwordieMS users.

    Using instance method SwordieDB::get_user_by_username(username) or the SwordieDB::get_char_by_name(name).user getter
    will create a User object instance with attributes identical to the user with username "username"
    (or IGN "name" for the latter) in the connected Swordie-based database.
    This class contains the appropriate getter and setter methods for said attributes.

    Attributes:
        account_type: Integer, representing the level of authority
        ban_reason: String (256 Char), representing the reason the user is banned
        donation_points: Integer, representing the amount of DP the user has
        maple_points: Integer, representing the amount of Maple Points the user has
        nx_prepaid: Integer, representing the amoung of NX Prepaid the user has
        user_id: Integer, representing the User ID
        vote_points: Integer, representing the amount of VP the user has
    """

    def __init__(self, user_stats, database_config):
        """Emulates how user object is handled server-sided

        Args:
            user_stats: dictionary of user attributes, formatted in SwordieMS style
            database_config: dictionary of protected attributes from a SwordieDB object
        """

        self._user_stats = user_stats
        self._database_config = database_config

        self._id = 0
        self._ban_reason = ""
        self._vote_points = 0
        self._donation_points = 0
        self._maple_points = 0
        self._nx_prepaid = 0
        self._account_type = 0
        self._username = ""
        self._password = ""
        self._pic = ""
        self.init_user_stats()

    def init_user_stats(self):
        """Given a dictionary of stats from Swordie's DB we add them to User object's attributes

        Runs near the end of User::__init__(user_stats, database_config).
        It assigns the user attributes in user_stats to their respective protected attributes belonging to
        the User object instance.
        """
        self._id = self._user_stats["id"]
        self._ban_reason = self._user_stats["banReason"]
        self._vote_points = self._user_stats["votepoints"]
        self._donation_points = self._user_stats["donationpoints"]
        self._maple_points = self._user_stats["maplePoints"]
        self._nx_prepaid = self._user_stats["nxPrepaid"]
        self._account_type = self._user_stats["accounttype"]
        self._username = self._user_stats["name"]
        self._password = self._user_stats["password"]
        self._pic = self._user_stats["pic"]

    @property
    def user_stats(self):
        return self._user_stats

    @property
    def database_config(self):
        return self._database_config

    @property
    def username(self):
        return self._username

    @username.setter
    def username(self, x):
        self._username = x
        self.set_stat_by_column("name", x)

    @property
    def password(self):
        return self._password

    @property
    def pic(self):
        return self._pic

    @pic.setter
    def pic(self, x):
        self._pic = x
        self.set_stat_by_column("pic", x)

    @property
    def user_id(self):
        return self._id

    @property
    def ban_reason(self):
        return self._ban_reason

    @ban_reason.setter
    def ban_reason(self, reason):
        if len(reason) < 255:
            self.set_stat_by_column("banReason", reason)
            self._ban_reason = reason
        else:
            print("Ban Reason was too long.")

    @property
    def vote_points(self):
        return self._vote_points

    @vote_points.setter
    def vote_points(self, amount):
        self.set_stat_by_column("votepoints", amount)
        self._vote_points = amount

    def add_vote_points(self, amount):
        """Adds the specified amount to the current VP count

        Args:
            amount: Int, representing the number of vote points to be added to the current count
        """
        new_vp = int(amount) + self.vote_points
        self.vote_points = new_vp

    @property
    def donation_points(self):
        return self._donation_points

    @donation_points.setter
    def donation_points(self, amount):
        self.set_stat_by_column("donationpoints", amount)
        self._donation_points = amount

    def add_donation_points(self, amount):
        """Adds the specified amount to the current DP count

        Args:
            amount: Int, representing the number of DPs to be added to the current count
        """
        new_dp = int(amount) + self.donation_points
        self.donation_points = new_dp

    @property
    def maple_points(self):
        return self._maple_points

    @maple_points.setter
    def maple_points(self, amount):
        self.set_stat_by_column("maplePoints", amount)
        self._maple_points = amount

    def add_maple_points(self, amount):
        """Adds the specified amount to the current Maple Points pool

        Args:
            amount: Int, representing the number of Maple Points to be added to the current pool
        """
        new_maple_points = int(amount) + self.maple_points
        self.maple_points = new_maple_points

    @property
    def account_type(self):
        return self._account_type

    @account_type.setter
    def account_type(self, acc_type):
        self.set_stat_by_column("accounttype", acc_type)
        self._account_type = acc_type

    def give_admin(self):
        """Sets the user account type to: Admin"""
        self.account_type = 4

    def give_intern(self):
        """Sets the user account type to: Intern"""
        self.account_type = 3

    def give_tester(self):
        """Sets the user account type to: Tester"""
        self.account_type = 5

    def give_player(self):
        """Sets the user account type to: Player"""
        self.account_type = 0

    def is_admin(self):
        """Checks if the user's account type is: Admin (Strictly)

        Checks if the account type ID associated with the user is 4. Both Int and String types are checked.

        Returns:
            Boolean, representing whether user is Admin (True) or not (False)
        """
        return self.account_type == 4 or self.account_type == "4"  # A string check just in case

    def change_password(self, new_pass):
        """WARNING: UNSAFE!

        Primitive password change function. It will break in real Swordie-based sources, as Swordie hashes passwords.
        Note: This is not a safe method to use when changing password as it does not hash the password after changing.

        Args:
            new_pass: string, representing the new plaintext password to be set to the database
        """
        self.set_stat_by_column("password", new_pass)

    def get_acc_type_string(self):
        """Fetches the account type of the user in words

        Fetches the account type (Int) from user attributes. Maps the account type to its meaning in words

        Returns:
            String, representing account type
        """
        if self.account_type == 4:
            return "Admin"
        elif self.account_type == 0:
            return "Player"
        elif self.account_type == 3:
            return "Intern"
        elif self.account_type == 5:
            return "Tester"

    @property
    def nx_prepaid(self):
        return self._nx_prepaid

    @nx_prepaid.setter
    def nx_prepaid(self, amount):
        self.set_stat_by_column("nxPrepaid", amount)
        self._nx_prepaid = amount

    def add_nx_prepaid(self, amount):
        """Adds the specified amount to the current NX pool

        Args:
            amount: Int, representing the amount of NX to be added to the NX pool
        """
        new_nx = int(amount) + self.nx_prepaid
        self.nx_prepaid = new_nx

    def get_stat_by_column(self, column):
        """Fetches user attribute by column name

        Returns:
            Int or String, representing user attribute queried
        Raises:
            Generic error on failure
        """
        try:
            return self.user_stats[str(column)]
        except Exception as e:
            print("[ERROR] Error trying to obtain the given column for table users.", e)

    def set_stat_by_column(self, column, value):
        """Sets a user's attributes by column name in database

        Grabs the database attributes provided through the class constructor.
        Uses these attributes to attempt a database connection.
        Attempts to update the field represented by the provided column in the users table, with the provided value.
        Not recommended to use this alone, as it won't update the user object which this was used from

        Args:
            value: int or string, representing the value to be set in the database
            column: string, representing the column in the database that is to be updated

        Returns:
            A boolean representing whether the operation was successful.

        Raises:
            SQL Error 2003: Can't cannect to DB
            WinError 10060: No response from DB
            List index out of range: Wrong column name
        """
        host = self._database_config["host"]
        user = self._database_config["user"]
        password = self._database_config["password"]
        schema = self._database_config["schema"]
        port = self._database_config["port"]

        try:
            database = con.connect(host=host, user=user, password=password, database=schema, port=port)

            cursor = database.cursor(dictionary=True)
            cursor.execute(f"UPDATE users SET {column} = '{value}' WHERE id = '{self.user_id}'")
            database.commit()
            print(f"Successfully updated {column} value for user id: {self.user_id}.")
            self._user_stats[column] = value  # Update the stats in the dictionary
            database.disconnect()
            return True
        except Exception as e:
            print("[ERROR] Error trying to set stats in database.", e)
            return False


swordie_db/database.py
"""SwordieDB is designed for use in development of SwordieMS-based MapleStory private server tools (e.g. Discord bots).

Copyright 2020 TEAM SPIRIT. All rights reserved.
Use of this source code is governed by a MIT-style license that can be found in the LICENSE file.
This module contains the main class that users would instantiate: SwordieDB.
Users can use this class to fetch and manipulate information from the database.
Refer to the project wiki on GitHub for more in-depth examples.

    Typical usage example:

    swordie = SwordieDB()  # Instantiate DB object
    char = swordie.get_char_by_name("brandon")  # Instantiate Character object
    meso = char.money  # Use of Character methods to fetch data from DB
    char.money = 123456789  # Use of Character methods to write data to DB
"""
import mysql.connector as con
from swordie_db.character import Character
from swordie_db.user import User


class SwordieDB:
    """Database object; models the SwordieMS DB.

    Use this class to create instances of SwordieMS characters, users, or inventories, complete with their respective
    data from the connected Swordie-based database.
    Using instance method SwordieDB::get_char_by_name("name") will create a Character object (see character.py) instance
    that has attributes identical to the character with IGN "name" in the connected Swordie-based database.

    Attributes:
        host: Optional; IP address of the database. Defaults to "localhost"
        schema: Optional; Name of the schema of the database (aka connection name). Defaults to "swordie"
        user: Optional; Username for access to the database. Defaults to "root"
        password: Optional; Password for access to the database. Defaults to ""
        port: Optional; Port with which to access the database. Defaults to 3306
    """
    def __init__(self, host="localhost", schema="swordie", user="root", password="", port=3306):
        self._host = host
        self._schema = schema
        self._user = user
        self._password = password
        self._port = port

        self._database_config = {
            "host": self.host,
            "user": self.user,
            "password": self.password,
            "schema": self.schema,
            "port": self.port
        }

    @property
    def host(self):
        return self._host

    @host.setter
    def host(self, x):
        self._host = x

    @property
    def schema(self):
        return self._schema

    @schema.setter
    def schema(self, x):
        self._schema = x

    @property
    def user(self):
        return self._user

    @user.setter
    def user(self, x):
        self._user = x

    @property
    def password(self):
        return self._password

    @password.setter
    def password(self, x):
        self._password = x

    @property
    def port(self):
        return self._port

    @port.setter
    def port(self, new_port):
        self._port = new_port

    def get_char_by_name(self, char_name):
        """Create an instance of a Character object from the given character name

        Uses the class constructor of the Character class to create a new instance, with the corresponding
        character data and database attributes from the connected database.

        Args:
            char_name: string, representing character name (aka IGN)

        Returns:
            Character object instantiated with corresponding data from the connected database.
            Defaults to None if the operation fails.

        Raises:
            Generic error on failure - handled by the Character::get_db() method
        """
        character_stats = Character.get_db(
            self._database_config,
            f"SELECT * FROM characterstats WHERE name = '{char_name}'"
        )  # Fetch first result because there should only be one character with that name

        character = Character(character_stats, self._database_config)
        return character

    def get_user_by_username(self, username):
        """Given a username (NOT IGN), create a new user object instance

        Fetches the user attributes from the database by querying for username.
        uses the User class constructor to create a new User object instance, with the said attributes.
        Useful for getting account information from accounts with no characters.

        Args:
            username: String, representing the username used for logging the user into game

        Returns:
            User object with attributes identical to its corresponding entry in the database

        Raises:
            Generic error on failure - handled by the Character::get_db() method
        """
        user_stats = Character.get_db(
            self._database_config,
            f"SELECT * FROM users WHERE name = '{username}'"
        )  # Fetch first result because there should only be one character with that name

        user = User(user_stats, self._database_config)
        return user

    def set_char_stat(self, name, column, value):
        """Given a character name and column name, change its value in the database

        Args:
            column: string, representing the column in the database
            name: string, representing the character name in the database
            value: string/int, representing the value that is to be updated in the corresponding field

        Returns:
            boolean, representing whether the operation completed successfully

        Raises:
            SQL Error 2003: Can't cannect to DB
            WinError 10060: No response from DB
            List index out of range: Wrong column or character name
        """
        try:
            database = con.connect(host=self.host, user=self.user, password=self.password, database=self.schema, port=self.port)
            cursor = database.cursor(dictionary=True)
            cursor.execute(f"UPDATE characterstats SET {column} = '{value}' WHERE name = '{name}'")
            database.commit()
            database.disconnect()
            print(f"Successfully set {name}'s stats in database.")
            return True
        except Exception as e:
            print("[ERROR] Error trying to update character stats in Database.", e)
            return False

    def get_user_id_by_name(self, char_name):
        """Given a character name, retrieve its corresponding user id from database

        Uses static method Character::get_user_id_by_name() for core logic

        Args:
            char_name: string, representing the character name in the database

        Returns:
            String, representing the user ID in the database
            Defaults to None if the operation fails.

        Raises:
            Errors are handled and thrown by Character::get_user_id_by_name()
            SQL Error 2003: Can't cannect to DB
            WinError 10060: No response from DB
            List index out of range: Wrong character name
        """
        return Character.get_user_id_by_name(self._database_config, char_name)
